query/query.go

```diff
@@ -0,0 +1,202 @@
+}
+
+// Create a new query
+func (r *GormQueryRepository) Create(ctx context.Context, q *Query) error {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "create"}, time.Now())
+	q.ID = uuid.NewV4()
+	if q.Creator == uuid.Nil {
+		return errors.NewBadParameterError("creator cannot be nil", q.Creator).Expected("valid user ID")
+	}
+	// Parse fields to make sure that query is valid
+	exp, _, err := search.ParseFilterString(ctx, q.Fields)
+	if err != nil || exp == nil {
+		log.Error(ctx, map[string]interface{}{
+			"space_id": q.SpaceID,
+			"fields":   q.Fields,
+		}, "unable to parse the query fields")
+		return err
+	}
+	err = r.db.Create(q).Error
+	if err != nil {
+		// combination of name and space ID should be unique
+		if gormsupport.IsUniqueViolation(err, "queries_title_space_id_creator_unique") {
+			log.Error(ctx, map[string]interface{}{
+				"err":      err,
+				"title":    q.Title,
+				"space_id": q.SpaceID,
+			}, "unable to create query because a query with same title already exists in the space by same creator")
+			return errors.NewDataConflictError(fmt.Sprintf("query already exists with title = %s , space_id = %s, creator = %s", q.Title, q.SpaceID, q.Creator))
+		}
+		log.Error(ctx, map[string]interface{}{}, "error adding Query: %s", err.Error())
+		return err
+	}
+	return nil
+}
+
+// List all queries in a space
+func (r *GormQueryRepository) List(ctx context.Context, spaceID uuid.UUID) ([]Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "list"}, time.Now())
+	var objs []Query
+	err := r.db.Where("space_id = ?", spaceID).Find(&objs).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return nil, err
+	}
+	return objs, nil
+}
+
+// ListByCreator all queries in a space by a creator
+func (r *GormQueryRepository) ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "listbycreator"}, time.Now())
+	var objs []Query
+	err := r.db.Where("space_id = ? AND creator=?", spaceID, creatorID).Find(&objs).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return nil, err
+	}
+	return objs, nil
+}
+
+// Load Query in a space
+func (r *GormQueryRepository) Load(ctx context.Context, ID uuid.UUID, spaceID uuid.UUID) (*Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "query", "show"}, time.Now())
+	q := Query{}
+	tx := r.db.Where("id = ? and space_id = ?", ID, spaceID).First(&q)
+	if tx.RecordNotFound() {
+		log.Error(ctx, map[string]interface{}{
+			"query_id": ID.String(),
+		}, "record not found")
+		return nil, errors.NewNotFoundError("query", ID.String())
+	}
+	if tx.Error != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err":      tx.Error,
+			"query_id": ID.String(),
+		}, "unable to load the query by ID")
+		return nil, errors.NewInternalError(ctx, tx.Error)
+	}
+	return &q, nil
+}
+
+// IsDuplicate checks if there is already a query with same content by same user
```

Comment:
@pranavgore09 why not simply insert and check for a duplicate with [IsUniqueViolation`](https://github.com/fabric8-services/fabric8-wit/blob/master/gormsupport/postgres.go#L38) ? If you want to have on-key-press-validation in the UI, you will need this function of course.

Reply:
Yes, @kwk I tried that. Our unique condition includes `deleted_at`.
Let's say that there is a record already present with title1, space1, and crreator1
If I am trying to create the same record, DB accepts it and creates a row and we get 201.
But I am testing data conflict error here https://github.com/fabric8-services/fabric8-wit/pull/1824/files#diff-5167477ec144512fe6c115235dae2ca8R162 and without calling `IsDuplicate` manually I do not get an error from DB.

I was expecting the behavior of code same as you were. So there is something wrong in design https://github.com/fabric8-services/fabric8-wit/pull/1824/files#diff-69e4ee5c520dbf4a8b435472706ab68dR10 or something wrong with my understanding.

Reply:
Hm, if the unique condition includes the `deleted_at` column you should be able to create query "foo", delete it and recreate it again without a name clash. Only if there exists a query foo that is not deleted, recreation should be prohibited. I'm checking the query.

Reply:
fixed `index` in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/bbbd0b4dd1d3d1fd9266c87bda23c26d5a7d67a7
removed `IsDuplicate` in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/ca299ba46d036f827341c44276bc0282ff75759a
thanks @kwk